### PR TITLE
Add IODA converters for mls omi ompsnm converters

### PR DESCRIPTION
**Pull request by karpob**
## Description
Adds IODA converters for Ozone measurements from Microwave Limb Sounder (MLS), Ozone Monitoring Instrument (OMI) and Ozone Mapping Profiler Suite-Nadir Mapper (OMPS-NM). Each script will generate a IODA formatted file for a provided synoptic time and path to raw hdf5 instrument datasets (which are read in using netCDF4).  (I'd add Kris Wargan on this for review, but I guess he got kicked out of JCSDA-internal).

### Issue(s) addressed
none

## Acceptance Criteria (Definition of Done)
IODA formatted data can be produced and read into JEDI for MLS, OMI and OMPS-NM. 

## Dependencies
none

## Impact
none

### CODE REVIEW

File: src/compo/ompsnm_o3_nc2ioda.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+#Global Dictionaries.
+locationKeyList = [
+    ("latitude", "float"),
+    ("longitude", "float"),
+    ("dateTime", "long"),
+]
+
+ 
+# dictionary to map things we're putting into ioda and taking out of instrument native format
+ioda2nc = {}
```
Comment (CoryMartin-NOAA) [inline src/compo/ompsnm_o3_nc2ioda.py:39]: same comment again

File: src/compo/ompsnm_o3_nc2ioda.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    if(len(rawFiles) == 0):
+        sys.exit("No Raw Files Found!!!")    
+    # get start and end times for qc/cropping data in MLS native time format (TAI seconds since Jan 1, 1993.)    
+    startTAI = ( ( cycle_time - timedelta(hours=3) ) - datetime(1993,1,1,0) ).total_seconds()
+    endTAI = ( ( cycle_time + timedelta(hours=3) ) - datetime(1993,1,1,0) ).total_seconds()
+
+    # Read in the O3 data in window 
+    o3 = ompsnm(rawFiles, startTAI, endTAI)
+
+    #for k in o3.outdata.keys():
```
Comment (CoryMartin-NOAA) [inline src/compo/ompsnm_o3_nc2ioda.py:269]: remove debug lines

File: src/compo/mls_o3_nc2ioda.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+
+#Global Dictionaries.
+locationKeyList = [
+    ("latitude", "float"),
+    ("longitude", "float"),
+    ("air_pressure", "float"),
+    ("dateTime", "integer"),
+]
+
+ioda2nc = {}
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/mls_o3_nc2ioda.py:37]: I'd rather not. Both ways are kinda ugly, but I prefer the way that is less awkward to type.

File: src/compo/mls_o3_nc2ioda.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+        #only output desired levels (lmin through lmax)
+        dd = {}
+        idx, = np.where( ( np.asarray(d['dateTime'])>=self.startTAI ) & ( np.asarray(d['dateTime'])<=self.endTAI ) )
+        dd['valKey'] = d['valKey'][idx,self.lmin:self.lmax+1]
+        dd['precision'] = d['precision'][idx,self.lmin:self.lmax+1]
+        lvec = np.arange(self.lmin+1,self.lmax+2)
+        dd['level'],dd['status'] = np.meshgrid(np.arange(self.lmin+1,self.lmax+2),d['status'][idx])
+
+        dd['air_pressure'],dd['dateTime'] = np.meshgrid(d['air_pressure'][self.lmin:self.lmax+1],d['dateTime'][idx])
+        _,dd['quality'] = np.meshgrid(lvec,d['quality'][idx])
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/mls_o3_nc2ioda.py:192]: I can replace these with np.tile, but I just used _ as a "tmp" which is disposed of anyway.

File: src/compo/mls_o3_nc2ioda.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        # if it is nrt, and the last file use all but first two and last 3 profiles
+        if( ifile == maxfile and self.nrt == True):
+            print("last file:{}".format(filename))
+            start = 2
+            end = end_of_file - 3
+        # otherwise if nrt skip first 2 and last 8 profiles to avoid duplicates.
+        elif( self.nrt == True ):
+            start = 2
+            end = end_of_file - 8
+        else:
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/mls_o3_nc2ioda.py:145]: Yes, it need to do different overlaps with MLS' weird overlap scheme.

File: src/compo/ompsnm_o3_nc2ioda.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+        self._setVarDict('integrated_layer_ozone_in_air')
+        self.outdata[self.varDict['integrated_layer_ozone_in_air']['valKey']] = []
+        #self.outdata[self.varDict['integrated_layer_ozone_in_air']['errKey']] = []
+
+        self._read()
+
+    # set ioda variable keys
+    def _setVarDict(self,iodavar):
+        self.varDict[iodavar]['valKey'] = iodavar, iconv.OvalName()
+        #self.varDict[iodavar]['errKey'] = iodavar, iconv.OerrName()
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/ompsnm_o3_nc2ioda.py:99]: We're going to set error in the yaml, and there are multiple qc_flags anway. Thought is to have them explicitly mapped to the data product.

File: src/compo/mls_o3_nc2ioda.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (CoryMartin-NOAA) [reply src/compo/mls_o3_nc2ioda.py:145]: I meant more of the fact that it is set before the if, and the else is equivalent

File: src/compo/mls_o3_nc2ioda.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (CoryMartin-NOAA) [reply src/compo/mls_o3_nc2ioda.py:37]: Ok as long as it passes coding norms, I'm happy

File: src/compo/mls_o3_nc2ioda.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (CoryMartin-NOAA) [reply src/compo/mls_o3_nc2ioda.py:192]: ok got it, just not very clear from a first pass

File: src/compo/mls_o3_nc2ioda.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/mls_o3_nc2ioda.py]: Ah, yeah. I removed the extraneous else statement.

File: src/compo/mls_o3_nc2ioda.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/mls_o3_nc2ioda.py]: I replaced all of those `_, variable = np.meshgrid`  with `np.tile` instead.

File: src/compo/mls_o3_nc2ioda.py [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+                print("Not Performing QC.")
+                d = self._just_flatten(nc_data)
+            for v in list(d.keys()):
+                if(v != 'valKey' and v != 'errKey'):
+                    self.outdata[(v,'MetaData')].extend(d[v]) 
+            for ncvar, iodavar in obsvars.items():
+                self.outdata[self.varDict[iodavar]['valKey']].extend(d['valKey'])
+                if(self.qcOn):
+                    self.outdata[self.varDict[iodavar]['errKey']].extend(d['errKey'])
+                #self.outdata[self.varDict[iodavar]['qcKey']] = qc_flag
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/mls_o3_nc2ioda.py]: deleted.

File: src/compo/ompsnm_o3_nc2ioda.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+        # unmask ground pixel quality to pass fill value.
+        d['ground_pixel_quality'].mask = False
+        # mesh time and scan_position to get flattened array instead of using loops
+        time_vec =  d['dateTime']
+        scan_position_vec = np.arange(1,d['valKey'].shape[1]+1)
+        d['scan_position'], d['dateTime'] = np.meshgrid(scan_position_vec,time_vec)
+        d['scan_position'] = d['scan_position'].astype('float32')
+        _,d['measurement_quality_flags'] = np.meshgrid(scan_position_vec,d['measurement_quality_flags'])
+        _,d['instrument_quality_flags'] = np.meshgrid(scan_position_vec, d['instrument_quality_flags'])
+        idx = np.where( (~d['valKey'].mask) )# & (d['dateTime']<=self.endTAI) & (d['dateTime']>=self.startTAI) )
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/ompsnm_o3_nc2ioda.py]: I put it back. It is needed. I commented it out for some debugging, but forgot about it.

File: src/compo/ompsnm_o3_nc2ioda.py (hunk 2 above)
Comment (karpob) [reply src/compo/ompsnm_o3_nc2ioda.py]: Removed

File: src/compo/ompsnm_o3_nc2ioda.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+        self.varAttrs[vkey]['units'] = 'Pa'
+
+    # Read data needed from raw MLS file.
+    def _read_nc(self,filename):
+        print("Reading: {}".format(filename))
+        d = {}
+        ncd = nc.Dataset(filename, 'r')
+        # use dictionary above to just read fields we want out of the netcdf.
+        for k in list(ioda2nc.keys()):
+            #print("Reading Field: {}"ioda2nc[k])
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/ompsnm_o3_nc2ioda.py]: Removed debug line.

File: src/compo/mls_o3_nc2ioda.py [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+        AttrData['nlocs'] = np.int32(DimDict['nlocs'])
+        # run a time duplicate check to see if NRT works.
+        for k in self.outdata.keys():
+            self.outdata[k] = np.asarray(self.outdata[k])
+            if(self.outdata[k].dtype =='float64'):
+                self.outdata[k] = self.outdata[k].astype('float32')
+            elif(self.outdata[k].dtype == 'int64' and k != ('dateTime','MetaData')):
+                self.outdata[k] = self.outdata[k].astype('int32') 
+
+        # EOS AURA uses TAI93 so add seconds offset from UNIX time for IODA
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/mls_o3_nc2ioda.py]: Nice, didn't know that! I changed to use 1993 Jan 1 as the epoch.  One less line of code sounds good to me.

File: src/compo/mls_o3_nc2ioda.py [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+    elif( args.hour != 0 and nrt ):
+        rawFiles = glob.glob( os.path.join(args.input, args.prefix+"*{}d".format(current_year)+current_doy+"*.he5") )
+    else:
+        rawFiles = glob.glob( os.path.join(args.input, args.prefix+"*{}d".format(current_year)+current_doy+".he5") )
+
+    rawFiles.sort()
+    if(nrt):
+        #limit files only between start and end of window.    
+        rawFilesOut = []
+        startDateWindow = cycle_time - timedelta(hours=3)
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/mls_o3_nc2ioda.py:351]: Yes, I added an optional arg to specify window length. The default is 6 hours.

File: src/compo/mls_o3_nc2ioda.py [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+    endTAI = ( ( cycle_time + timedelta(hours=3) ) - datetime(1993,1,1,0) ).total_seconds()
+
+
+    # Read in the O3 data in window over selected levels (if not default 8-49)
+    # if nrt is set, assumes near real-time and will skip first two and last 8 profiles for most files 
+    # (last 3 profiles skipped if it is the last file in window)
+    # RTM regarding Near Real time (NRT) in 
+    # https://discnrt1.gesdisc.eosdis.nasa.gov/data/Aura_NRT/ML2SO2_NRT.005/doc/NRT-user-guide-v5.pdf
+    o3 = mls(rawFiles, lmin-1, lmax-1, startTAI, endTAI, nrt,args.qc)
+    #for k in o3.outdata.keys():
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/mls_o3_nc2ioda.py]: Removed commented out debug prints.

File: src/compo/omi_o3_nc2ioda.py [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+    return (integer_value & (1 << bit_position)) != 0
+
+#Global Dictionaries.
+locationKeyList = [
+    ("latitude", "float"),
+    ("longitude", "float"),
+    ("dateTime", "integer"),
+]
+
+ioda2nc = {}
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/omi_o3_nc2ioda.py:40]: Leaving it as-is as mentioned earlier.

File: src/compo/omi_o3_nc2ioda.py [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+        for v in vars2output:
+            if(v != 'valKey'):
+                self.outdata[(v, 'MetaData')] = []
+        self.outdata[self.varDict['integrated_layer_ozone_in_air']['valKey']] = []
+        self._read()
+
+    # set ioda variable keys
+    def _setVarDict(self,iodavar):
+        self.varDict[iodavar]['valKey'] = iodavar, iconv.OvalName()
+        #self.varDict[iodavar]['errKey'] = iodavar, iconv.OerrName()
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/omi_o3_nc2ioda.py]: Yeah, again, going to do QC/error in the yaml, and there are multiple flags, so it's easier (I think) to just map to the data product and not add another layer of mapping to ioda's qc flag.

[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/omi_o3_nc2ioda.py]: Removed to commented out line though..

File: src/compo/omi_o3_nc2ioda.py [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+            dd[v] = []
+ 
+        for itime in range(d['latitude'].shape[0]):
+            for iscan in range(d['latitude'].shape[1]):
+                if(d['dateTime'][itime] < self.startTAI or d['dateTime'][itime] > self.endTAI):
+                    continue
+                if (d['prior_o3'][itime,iscan, 0] <= 0.0 or d['valKey'][itime, iscan] <= 0.0):
+                    continue
+                # from Kris' fortran 
+                #!! A hack to use the quality flags as recommended by the OMI team (Jul 2019)
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/omi_o3_nc2ioda.py]: It's actually not a hack. The comment that I extracted from Kris' code says that, but it's refers to switching a flag to a value the GSI will flag after an update to the suggested QC from the OMI instrument team. I changed the comment to reflect that.

File: src/compo/omi_o3_nc2ioda.py [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+                print('Not doing QC.')
+                d = self._just_flatten(nc_data)
+            # add MetaData variables.
+            for v in list(d.keys()):
+                if(v != 'valKey' ):
+                    self.outdata[(v, 'MetaData')].extend(d[v])
+            
+            for ncvar, iodavar in obsvars.items():
+                self.outdata[self.varDict[iodavar]['valKey']].extend(d['valKey'])
+                #self.outdata[self.varDict[iodavar]['qcKey']].extend(d['qcKey'])
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/omi_o3_nc2ioda.py]: Removed extra line. Pushing QC/error into downstream yaml

File: src/compo/ompsnm_o3_nc2ioda.py (hunk 1 above)
Comment (karpob) [reply src/compo/ompsnm_o3_nc2ioda.py:38]: Again, keeping the dictionary approach from before.

File: src/compo/ompsnm_o3_nc2ioda.py [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+        vars2output.append('scan_position')
+        for v in vars2output:
+            if(v != 'valKey'):
+                self.outdata[(v, 'MetaData')] = []
+        self.outdata[self.varDict['integrated_layer_ozone_in_air']['valKey']] = []
+
+ 
+        self._setVarDict('integrated_layer_ozone_in_air')
+        self.outdata[self.varDict['integrated_layer_ozone_in_air']['valKey']] = []
+        #self.outdata[self.varDict['integrated_layer_ozone_in_air']['errKey']] = []
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/ompsnm_o3_nc2ioda.py]: Yeah, again, going to do QC/error in the yaml, and there are multiple flags, so it's easier (I think) to just map to the data product and not add another layer of mapping to ioda's qc flag.

File: src/compo/ompsnm_o3_nc2ioda.py [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+            #print("Reading Field: {}"ioda2nc[k])
+            d[k] = ncd[ ioda2nc[k] ][...]
+        # unmask ground pixel quality to pass fill value.
+        d['ground_pixel_quality'].mask = False
+        # mesh time and scan_position to get flattened array instead of using loops
+        time_vec =  d['dateTime']
+        scan_position_vec = np.arange(1,d['valKey'].shape[1]+1)
+        d['scan_position'], d['dateTime'] = np.meshgrid(scan_position_vec,time_vec)
+        d['scan_position'] = d['scan_position'].astype('float32')
+        _,d['measurement_quality_flags'] = np.meshgrid(scan_position_vec,d['measurement_quality_flags'])
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/ompsnm_o3_nc2ioda.py]: replaced them with np.tile as done earlier.

File: src/compo/mls_o3_nc2ioda.py [hunk 19]
```diff
@@ -0,0 +1,400 @@
...
+        return dd
+                
+    def _read(self):
+        # set up variable names for IODA
+        self._setVarAttr('mole_fraction_of_ozone_in_air')
+
+       # loop through input filenames
+        for i,f in enumerate(self.filenames):
+            nc_data = self._read_nc(f,i,len(self.filenames)-1)
+            #dups = [item for item, count in Counter(time).items() if count > 1]
```
[the remark this answers is not in the document]
Comment (karpob) [reply src/compo/mls_o3_nc2ioda.py]: deleted.

**Review by jeromebarre — CHANGES_REQUESTED** [review]
Thank for the effort of coding all this. Ideally it's better to isse a PR per instrument but for now that's fine.

We would need to have the ctest added with the new converters before merging. See examples in https://github.com/JCSDA-internal/ioda-converters/tree/develop/test

It not a lot of work to add and it ensures that we reviewers can test your code.

One minor thing is the error specification when qcOn is enabled for MLS. _calc_error is called in _do_qc. Should those ad-hoc errors specifications decoupled from the qc filtering? Meaning we could have a separate option for replacing the errors?

Thanks a lot

File: src/compo/mls_o3_nc2ioda.py [hunk 20]
```diff
@@ -0,0 +1,381 @@
...
+            ooe = ooe + (0.30 * abs(o3))  # 0.3 is my conservative guess
+        elif (lev == 8):  # 215 hPa
+            ooe = ooe + (0.20 * abs(o3))
+        elif (lev == 9):  # 177 hPa
+            ooe = ooe + (0.125 * abs(o3))
+        elif (lev == 10 or lev == 11 or lev == 12):  # 150-100 hPa
+            ooe = ooe + (0.05 * abs(o3))
+
+        ooe = np.sqrt(max((0.5*ooe)**2+(o3_prec)**2, 1.e-6))
+        return ooe
```
Comment (jeromebarre) [inline src/compo/mls_o3_nc2ioda.py:171]: This error specification is enabled for MLS when qcOn. _calc_error is called in _do_qc. Should those ad-hoc errors specifications decoupled from the qc filtering? Meaning we could have a separate option for replacing the errors?

Comment (karpob) [reply src/compo/mls_o3_nc2ioda.py:171]: I mean, I guess? But the two options I was thinking were do the QC and error calculation downstream or upstream. I don't know that we're that motivated to push more QC/error calculation within the ufo, so I'm dubious we'd use anything but the default.

Comment (jeromebarre) [reply src/compo/mls_o3_nc2ioda.py:171]: To clarify I would see it more flexible to have the option to do the pre qc filtering and not the ad-hoc errors and conversely. For that _calc_error should be called outside of _do_qc within a condition triggered by an option. It is more a suggestion than a requirement here. I am not completely new to MLS but it's been more than 10 years now... I remember to have to correct those errors for DA.

Comment (karpob) [reply src/compo/mls_o3_nc2ioda.py:172]: Ok, I broke it out, and moved it out of _do_qc.

**Review by jeromebarre — APPROVED** [review]
Looks good to me now. Thanks for this!

@CoryMartin-NOAA are you satisfied approve this for merge?

**Review by CoryMartin-NOAA — APPROVED** [review]

**Pull request merged**